packages/logger/src/logger.ts

```diff
@@ -0,0 +1,147 @@
+/**
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * A container for all of the Logger instances
+ */
+export const instances: Logger[] = [];
+
+/**
+ * The JS SDK supports 5 log levels and also allows a user the ability to
+ * silence the logs altogether.
+ *
+ * The order is a follows:
+ * DEBUG < LOG < INFO < WARN < ERROR < SILENT
+ *
+ * All of the log types above the current log level will be captured (i.e. if
+ * I set the log level to `INFO`, errors will still be logged, but `DEBUG` and
+ * `LOG` logs will not)
+ */
+export enum LogLevel {
+  DEBUG,
+  LOG,
+  INFO,
+  WARN,
+  ERROR,
+  SILENT
+}
+
+/**
+ * A container for the default log level
+ */
+const defaultLogLevel: LogLevel = LogLevel.WARN;
+
+/**
+ * We allow users the ability to pass their own log handler. We will pass the
+ * type of log, the current log level, and any other arguments passed (i.e. the
+ * messages that the user wants to log) to this function.
+ */
+export type LogHandler = (
+  loggerInstance: Logger,
+  logType: LogLevel,
+  ...args: any[]
+) => void;
+
+/**
+ * The default log handler will forward DEBUG, LOG, INFO, WARN, and ERROR
+ * messages on to their corresponding console counterparts (if the log method
+ * is supported by the current log level)
+ */
+const defaultLogHandler: LogHandler = (instance, logType, ...args) => {
+  if (logType < instance.logLevel) return;
+  const now = new Date();
+  switch (logType) {
+    case LogLevel.SILENT:
+      return;
+    case LogLevel.LOG:
+      console.log(`[${now}]  ${instance.name}:`, ...args);
+      break;
+    case LogLevel.INFO:
+      console.info(`[${now}]  ${instance.name}:`, ...args);
+      break;
+    case LogLevel.WARN:
+      console.warn(`[${now}]  ${instance.name}:`, ...args);
+      break;
+    case LogLevel.ERROR:
+      console.error(`[${now}]  ${instance.name}:`, ...args);
+      break;
+    default:
```

Comment:
1) You left out DEBUG
2) Rather than default to console.log(), I'd throw an exception. That way you'll notice if you leave out a log level.  If for some reason you were intentionally handling DEBUG via the `default:` case, then a comment is warranted to explain your intent.

Reply:
Done